### ISSUE TITLE
 Issue with Handling Keyless Entities in Microsoft Dynamics OData V4 API

### ISSUE THREAD

**Issue by erik-kers** [odata2ts/odata2ts#241]
**Description:**
I am using `odata2ts` to interact with the Microsoft Dynamics API, which is based on OData V4. This API includes certain keyless entities, particularly some that are abstract and others that are open types. However, I'm encountering issues with these specific entities. The entities in question are:

- Abstract Entities: `crmbaseentity` and `crmmodelbaseentity`
- Open Type: `expando`

According to the OData V4 specification, these entities do not require keys. The reasons are as follows:

- **Abstract Entities (`crmbaseentity` and `crmmodelbaseentity`)**: As per OData V4, abstract entities are used as base types for other entities and are not meant to be instantiated directly. Hence, they do not require keys. Reference: [crmbaseentity](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/crmbaseentity?view=dataverse-latest), [crmmodelbaseentity](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/crmmodelbaseentity?view=dataverse-latest), [Abstract entity types (Microsoft)](https://learn.microsoft.com/en-us/odata/webapi/abstract-entity-types), [Edm.Untyped](https://docs.oasis-open.org/odata/new-in-odata/v4.01/cn04/new-in-odata-v4.01-cn04.html#sec_NewBuiltinAbstractTypeEdmUntyped), [Abstrac Entity type](https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_AbstractEntityType)
- **Open Type (`expando`)**: Open types in OData V4 allow for dynamic properties that are not defined in the model. Given their dynamic and flexible nature, they may not have keys defined. Reference: [expando](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/expando?view=dataverse-latest).

**Issue:**
When trying to work with these entities using `odata2ts`, I am facing:

```bash
Loaded config file:  odata2ts.config.ts
---------------------------
Starting generation process
Found metadata file at:  resource/keyless.xml
Reading metadata from file: resource/keyless.xml
Error while running the program Error: Key property is missing from Entity "AbstractBaseEntity" (AbstractBaseEntity)!
    at C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\@odata2ts\odata2ts\lib\data-model\DataModelDigestion.js:229:23
    at Array.forEach (<anonymous>)
    at DigesterV4.postProcessModel (C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\@odata2ts\odata2ts\lib\data-model\DataModelDigestion.js:218:20)
    at DigesterV4.digestEntityTypesAndOperations (C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\@odata2ts\odata2ts\lib\data-model\DataModelDigestion.js:137:14)
    at DigesterV4.<anonymous> (C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\@odata2ts\odata2ts\lib\data-model\DataModelDigestion.js:106:18)
    at Generator.next (<anonymous>)
    at C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\tslib\tslib.js:169:75
    at new Promise (<anonymous>)
    at Object.__awaiter (C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\tslib\tslib.js:165:16)
    at DigesterV4.digest (C:\Users\erik.kers\repositories\Privat\odata2ts-error-reproduction\node_modules\@odata2ts\odata2ts\lib\data-model\DataModelDigestion.js:105:24)
```

 "Error while running the program Error: Key property is missing from Entity "EmptyBaseEntity" (EmptyBaseEntity)!". This seems to stem from the library's handling (or lack thereof) of keyless entities, particularly in the context of abstract and open types.

**Question:**
Can `odata2ts` be able to handle keyless entities, including abstract and open types, as per the OData V4 specification?
Reading the documentation I haven't found a solution for my problem. I have tried, hackily, to assign both keys and properties to these types of entities in the postProcess-stage. It has worked up until odata2ts tries to generate the odata service. But the "solution" felt quite hacky.

I'm quite new to odata so I might also have misunderstood something, in that case I'm sorry.

**Steps to Reproduce:**
1. Create or get a odata schema with entity types that are abstract or open type and has no keys
2. Run `odata2ts` as you would normally do
3. You will get an error because i.e. the abstract entity type did not have a key
   
**Additional Context:**
I created a repo reproducing the issue here: https://github.com/erik-kers/odata2ts-error-reproduction

**Comment by texttechne** [comment]
Hi @erik-kers,

great detailed issue with sample repo! Kudos!

Ok, abstract entities shouldn't be that hard to accomodate for. Derived entity types work as demonstrated by your example with keys, so it's more like allowing keyless entity types. I just need to find the time. 

I've prepared odata2ts for `open types` in the sense, that request and response may have additional properties. You won't get any exception and props are not filtered out or something. Of course, you have to take care of the typing yourself at that point by creatíng your own types based on the generated types... without metadata there's nothing odata2ts can do in this regard.

However, I have problems to understand how `expandos` should work without keys when they are declared as `EntityType`. Hence, I think that this example is simply wrong:
```xml
              <EntityType Name="OpenBaseEntity" OpenType="true" />

              <EntityType Name="DynamicObject" BaseType="ExampleModel.OpenBaseEntity">
                <Key>
                    <PropertyRef Name="Id" />
                </Key>
                <Property Name="Model" Type="Edm.String" />
                <Property Name="Year" Type="Edm.Int32" />
            </EntityType>
```
The main difference between `EntityType` and `ComplexType` is that the former requires a key while the latter does not. Only by declaring the `EntityType` as `abstract` you don't have to specify the key. See [this explanation by Microsoft](https://learn.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-v4/use-open-types-in-odata-v4):
> First, some OData terminology:
>
>Entity type: A structured type with a key.
>Complex type: A structured type without a key.
>Open type: A type with dynamic properties. Both entity types and complex types can be open.

Imagine you would expose the `OpenBaseEntity` as `EntitySet`. You won't be able to address a specific entity by id, because the key isn't known: `OpenBaseEntity(???)`. The key must be known at compile time and cannot be used dynamically at runtime. 

Additional reading: [Open Types explained by odata.org](https://www.odata.org/getting-started/advanced-tutorial/#openType)

So this issue boils down to allowing `abstract types`, right? Or do you have a specific problem with `expandos`?

**Comment by erik-kers** [comment]
Hi @texttechne 

I'm sorry about the late reply. I've been trying to get ahold of a real metadata-document but it doesn't seem like I will be able to do this.

You are correct in that I've made an error regarding the OpenType. I will instead give you the real example with expando.

```xml
<Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="Microsoft.Dynamics.CRM" Alias="mscrm">
<EntityType Name="crmbaseentity" Abstract="true"/>
<EntityType Name="expando" BaseType="mscrm.crmbaseentity" OpenType="true"/>
```

As you can see it has tyhe crmbaseentity as it's base type. And there are no other types using it as the base type. I don't know how I got that messed up tbh.

This is for example the resulting Model.d.ts I get when I give each keyless entity a faux-key
```ts
export interface Icrmbaseentity {
    __key_DO_NOT_USE: string | null;
}
export type IcrmbaseentityId = string | {
    __key_DO_NOT_USE: string | null;
};
export interface IEditablecrmbaseentity {
}
export interface Iexpando extends Icrmbaseentity {
}
```

I understand this is because of 

> I've prepared odata2ts for open types in the sense,

So in summary, yes. the issue boils down to allowing types.

**Comment by nickcomua** [comment]
I'm also tried this for Microsoft Datavers. Had same issue.

```
Starting generation process
Found metadata file at:  resource/trippin.xml   
Reading metadata from file: resource/trippin.xml
Error while running the program Error: Key property is missing from Entity "crmbaseentity" (crmbaseentity)!
```

**Comment by texttechne** [comment]
Hi @erik-kers & @nickcomua,

open types and abstract types - in fact keyless EntityTypes - are now supported by the generator in version `0.33.0`.

@erik-kers please test and close this issue if it works.

**Comment by erik-kers** [comment]
Hi @texttechne!

Awesome release - it looks as if it did solve the issue regarding abstract entities! However I'm now getting another problem some while into the generation.

```
Error while running the program NotImplementedError: Unexpected! Inserting syntax kind of InterfaceDeclaration, but ExpressionStatement was inserted.
    at getRangeWithoutCommentsFromArray (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:917:23)
    at insertIntoBracesOrSourceFileWithGetChildren (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:2523:12)
    at SourceFile._insertChildren (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:11489:20)
    at SourceFile.insertInterfaces (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:11301:25)
    at SourceFile.addInterfaces (C:\Users\ererik-kerskk\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:11295:25)
    at SourceFile.addInterface (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\ts-morph\dist\ts-morph.js:11292:25)
    at ModelGenerator.generateModel (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\@odata2ts\odata2ts\lib\generator\ModelGenerator.js:59:25)
    at C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\@odata2ts\odata2ts\lib\generator\ModelGenerator.js:39:18
    at Array.forEach (<anonymous>)
    at ModelGenerator.generateModels (C:\Users\erik-kers\source\repos\odata2ts-error-reproduction-no-git\node_modules\@odata2ts\odata2ts\lib\generator\ModelGenerator.js:38:41)
```

Perhaps @nickcomua could reproduce it also?

I will have to look into this and see if I can provide any more information and perhaps try to figure out when it crashes. 

In any case, perhaps this issue should be closed since this might not neccessarily be related to the original issue, what do you think @texttechne?

**Comment by texttechne** [comment]
Hi @erik-kers,

please provide the metadata file in question. If it's sensible info, then private repo and you invite me.

> perhaps this issue should be closed since this might not neccessarily be related to the original issue

Perhaps. Let's first figure out what it is 😄, if different, would be cool if you open another issue...
but for now, I would love to get hold of the example...

**Comment by erik-kers** [comment]
@texttechne  I'll see to it.

**Comment by texttechne** [comment]
Ok, weired error, but the problem is much bigger. I'll create another issue and close this as you have tested this bug.

**Comment by texttechne** [comment]
See #251
